File: java-selenium-junit5-cucumber/src/test/java/com/saucedemo/uitests/pages/InventoryPage.java
package com.saucedemo.uitests.pages;

import java.time.Duration;
import java.util.List;
import org.openqa.selenium.By;
import org.openqa.selenium.NoSuchElementException;
import org.openqa.selenium.StaleElementReferenceException;
import org.openqa.selenium.TimeoutException;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;

public final class InventoryPage extends BasePage {

    private static final By PRODUCT_NAMES = By.cssSelector("[data-test='inventory-item-name']");
    private static final By CART_BADGE = By.cssSelector("[data-test='shopping-cart-badge']");
    private static final By CART_LINK = By.cssSelector("[data-test='shopping-cart-link']");

    private static final Duration CONFIRM_TIMEOUT = Duration.ofSeconds(3);
    private static final int MAX_TOGGLE_ATTEMPTS = 3;

    public InventoryPage(WebDriver driver) {
        super(driver);
    }

    /**
     * XPath by CSS class plus visible button text. Verified against the live DOM captured by a failing CI
     * run in this gallery's Selenium siblings: each product sits in a {@code <div class="inventory_item">}
     * containing {@code <div data-test="inventory-item-name">}, and the toggle is
     * {@code <button>Add to cart</button>} / {@code <button>Remove</button>}.
     *
     * <p>(saucedemo does also expose {@code data-test="add-to-cart-<slug>"} on those buttons, so a
     * slug-based locator would work too; this one is kept because it needs no slugification.)
     */
    private static By addToCartButtonFor(String productName) {
        return By.xpath("//div[@class='inventory_item'][.//div[@data-test='inventory-item-name' and text()='"
                + productName + "']]//button[text()='Add to cart']");
    }

    private static By removeFromCartButtonFor(String productName) {
        return By.xpath("//div[@class='inventory_item'][.//div[@data-test='inventory-item-name' and text()='"
                + productName + "']]//button[text()='Remove']");
    }

    public boolean isDisplayed() {
        return waitAndCheckVisible(PRODUCT_NAMES);
    }

    public List<String> listProductNames() {
        return driver.findElements(PRODUCT_NAMES).stream().map(WebElement::getText).toList();
    }

    public void addToCart(String productName) {
        clickAndConfirmToggle(
                addToCartButtonFor(productName), removeFromCartButtonFor(productName), productName, "added to");
    }

    public void removeFromCart(String productName) {
        clickAndConfirmToggle(
                removeFromCartButtonFor(productName), addToCartButtonFor(productName), productName, "removed from");
    }

    /**
     * Establishes an empty cart as a scenario precondition, and (because it waits for the product grid
     * first) doubles as the "login has finished rendering" barrier that {@link LoginPage#submitLogin} does
     * not provide on its own.
     *
     * <p>In practice the removal loop is normally a no-op: saucedemo keeps the cart client-side and every
     * scenario gets a brand-new browser profile (see {@code WebDriverFactory}'s per-session
     * --user-data-dir), so nothing carries over between scenarios or runs. It is kept as an explicit,
     * cheap guard so a scenario asserting exact cart contents can never inherit unexpected state.
     */
    public void clearCart() {
        // listProductNames() is an instant, non-waiting read, so calling it as the very first post-login
        // action risks reading a not-yet-hydrated page and silently clearing nothing. Waiting for the
        // product grid here is what login itself does not guarantee.
        waitForVisible(PRODUCT_NAMES);

        for (String productName : listProductNames()) {
            if (isVisible(removeFromCartButtonFor(productName))) {
                removeFromCart(productName);
            }
        }
    }

    /**
     * The number on the cart badge, or 0 when no badge is rendered - which is how saucedemo shows an empty
     * cart. Reads the element exactly once and never throws: callers poll this repeatedly, so a
     * check-then-read pair would leave a window where the badge disappears between the two calls and the
     * read then blocks for the full explicit wait before throwing - failing the assertion instead of
     * simply reporting 0.
     */
    public int cartCount() {
        try {
            WebElement badge = driver.findElement(CART_BADGE);
            if (!badge.isDisplayed()) {
                return 0;
            }
            try {
                return Integer.parseInt(badge.getText().trim());
            } catch (NumberFormatException e) {
                return 0;
            }
        } catch (NoSuchElementException | StaleElementReferenceException e) {
            return 0;
        }
    }

    public CartPage openCart() {
        click(CART_LINK);
        return new CartPage(driver);
    }

    /**
     * Clicks a toggle button (add-to-cart / remove) and confirms the click actually took effect by waiting
     * for its counterpart button to appear, retrying a few times if it does not.
     *
     * <p>Observed directly in CI in this gallery's Selenium siblings: an "Add to cart" click that reported
     * success (no exception, well under a second) was followed by the cart badge reading 0 for a full
     * 5-second poll - the click reached a real, visible, enabled button, but the app's cart state never
     * changed. A plausible cause is a brief window where the button is visibly interactive before its
     * click handler is fully wired up (a client-side hydration race); re-clicking after confirming the
     * expected effect did not happen is the general-purpose fix regardless of the exact mechanism.
     */
    private void clickAndConfirmToggle(By clickLocator, By counterpartLocator, String productName, String action) {
        int clicks = 0;

        // The button must exist before any of this makes sense. Waiting for it here (rather than relying
        // on the instant isVisible check below) means a not-yet-rendered grid is waited out instead of
        // being mistaken for "the click didn't take effect".
        waitForClickable(clickLocator);

        for (int attempt = 1; attempt <= MAX_TOGGLE_ATTEMPTS; attempt++) {
            // Only re-click while the toggle is still in its "before" state - if an earlier attempt did
            // land (just slower than the confirm timeout), the button will already have flipped, and
            // re-clicking would mean waiting out click()'s full explicit-wait timeout for nothing.
            if (isVisible(clickLocator)) {
                click(clickLocator);
                clicks++;
            }

            if (waitAndCheckVisible(counterpartLocator, CONFIRM_TIMEOUT)) {
                return;
            }
        }

        // Report the real click count. Saying "clicked 3 times" when the button was never in a clickable
        // state (so nothing was clicked at all) points debugging at the app instead of at the page state,
        // which is exactly backwards.
        throw new TimeoutException("'" + productName + "' was never actually " + action + " the cart: over "
                + MAX_TOGGLE_ATTEMPTS + " attempts the toggle button was clicked " + clicks
                + " time(s) and its counterpart (" + counterpartLocator + ") never appeared.");
    }
}

File: ts-playwright-test/src/pages/inventoryPage.ts
import { expect } from '@playwright/test';
import type { Locator, Page } from '@playwright/test';
import { BasePage } from './basePage';
import { CartPage } from './cartPage';

export class InventoryPage extends BasePage {
  /** All six product-name cells on the grid. Tests assert on `.count()` / `allTextContents()` via web-first assertions. */
  readonly productNames: Locator;
  /** Absent from the DOM entirely when the cart is empty - `await expect(cartBadge).toBeHidden()` is the empty-cart assertion. */
  readonly cartBadge: Locator;
  private readonly cartLink: Locator;

  constructor(page: Page) {
    super(page);
    this.productNames = page.getByTestId('inventory-item-name');
    this.cartBadge = page.getByTestId('shopping-cart-badge');
    this.cartLink = page.getByTestId('shopping-cart-link');
  }

  /**
   * One product tile, found by filtering the grid down to the tile containing an exact-text match for
   * the product name.
   *
   * This replaces the interpolated XPath the three sibling frameworks share
   * (`//div[@class='inventory_item'][.//div[@data-test='inventory-item-name' and text()='<name>']]//button[text()='Add to cart']`).
   * The two gains are worth being explicit about: string-interpolating a product name into an XPath
   * predicate breaks on any name containing a quote, and matching a button by `text()` in XPath is a
   * structural match that says nothing about the element being a button at all. `getByRole('button')`
   * below asserts the accessible role, which is both what a user perceives and what survives markup
   * changes that leave the role intact.
   *
   * The `.inventory_item` container class is the one DOM fact carried over unchanged - it was verified
   * against the live saucedemo DOM during the Selenium sibling's CI debugging, so it is not a guess.
   */
  private itemFor(productName: string): Locator {
    return this.page
      .locator('.inventory_item')
      .filter({ has: this.page.getByText(productName, { exact: true }) });
  }

  // `exact: true` because getByRole's accessible-name matching is substring-based by default. The two
  // saucedemo button labels do not overlap, so substring matching would work today - but an exact match
  // is what the sibling frameworks' verified XPath predicate (`//button[text()='Add to cart']`) actually
  // asserts, so this keeps the migrated locator equivalent rather than looser.
  private addButtonFor(productName: string): Locator {
    return this.itemFor(productName).getByRole('button', { name: 'Add to cart', exact: true });
  }

  private removeButtonFor(productName: string): Locator {
    return this.itemFor(productName).getByRole('button', { name: 'Remove', exact: true });
  }

  async addToCart(productName: string): Promise<void> {
    await this.toggle(
      this.addButtonFor(productName),
      this.removeButtonFor(productName),
      productName,
      'added to'
    );
  }

  async removeFromCart(productName: string): Promise<void> {
    await this.toggle(
      this.removeButtonFor(productName),
      this.addButtonFor(productName),
      productName,
      'removed from'
    );
  }

  /**
   * Establishes an empty cart as a test precondition, and (because it waits for the product grid first)
   * doubles as the "login has finished rendering" barrier that `LoginPage.submitLogin` does not provide.
   *
   * Normally a no-op: saucedemo keeps the cart client-side and every test gets a fresh, isolated
   * `BrowserContext`, so nothing carries over. Kept as a cheap explicit guard so a test asserting exact
   * cart contents can never inherit unexpected state.
   */
  async clearCart(): Promise<void> {
    await expect(this.productNames.first()).toBeVisible();

    for (const productName of await this.productNames.allTextContents()) {
      if (await this.removeButtonFor(productName).isVisible()) {
        await this.removeFromCart(productName);
      }
    }
  }

  async openCart(): Promise<CartPage> {
    await this.cartLink.click();
    return new CartPage(this.page);
  }

  /**
   * Clicks a toggle button (add-to-cart / remove) and confirms the click took effect by waiting for its
   * counterpart to appear, re-clicking a few times if not.
   *
   * Retained deliberately from the sibling frameworks rather than dropped as redundant. Playwright's
   * actionability checks do make a lost click much less likely than Selenium's, but the Selenium
   * sibling's own investigation concluded its residual CI flakiness was most likely saucedemo.com
   * degrading under repeated automated traffic - and no amount of client-side waiting fixes a server
   * that never changes state. This is the one hand-rolled retry in this framework, and it retries an
   * *action*, which is why a web-first assertion cannot replace it: `expect().toBeVisible()` re-checks,
   * it does not re-click.
   *
   * `isVisible()` (instant, non-retrying) is the right check on later attempts: if an earlier click
   * actually landed, the button has already flipped, and re-clicking would mean waiting out a full
   * timeout for nothing.
   */
  private async toggle(
    clickTarget: Locator,
    counterpart: Locator,
    productName: string,
    action: string
  ): Promise<void> {
    const maxAttempts = 3;
    const confirmTimeoutMs = 3000;
    let clicks = 0;

    for (let attempt = 1; attempt <= maxAttempts; attempt++) {
      if (attempt === 1 || (await clickTarget.isVisible())) {
        await clickTarget.click();
        clicks++;
      }

      try {
        await counterpart.waitFor({ state: 'visible', timeout: confirmTimeoutMs });
        return;
      } catch {
        // Not yet - fall through to the next attempt. A genuinely ambiguous locator would have thrown a
        // strict-mode violation on the click above, before ever reaching this wait.
      }
    }

    // Report the real click count. Saying "clicked 3 times" when the button was never in a clickable
    // state (so nothing was clicked at all) points debugging at the app instead of at the page state,
    // which is exactly backwards.
    throw new Error(
      `'${productName}' was never actually ${action} the cart: over ${maxAttempts} attempts the toggle ` +
        `button was clicked ${clicks} time(s) and its counterpart never appeared.`
    );
  }
}
